### DISCUSSION TITLE
Help needed on conversion to object: json.hpp:5527:67: error: incomplete type ... used in nested name specifier

### DISCUSSION THREAD

**Discussion by goofy2k** [nlohmann/json#3286] [Q&A] [answered]
Hi, I am a new user of this lib. I use json.hpp version 3.10.5.

I have an object user_dspPresets of type PresetsBank_t which is a map with int keys and values being vectors of float:
`
typedef vector<float> DSPpresetVector_t;
typedef map <int, DSPpresetVector_t> PresetsBank_t;
PresetsBank_t user_dspPresets;
`

After population of the object, I want to save it in a file. I decided to convert it to JSON, save that as a text file, for later retrieval.
After conversion to JSON I want to check the reverse conversion  to the same object type. 

code:

`  
   
    //convert to json
    //for arbitrary types you need to add a to_json and a from_json method, see: https://json.nlohmann.me/features/arbitrary_types/
    //for now we do it without those and test if that is OK 

    json j =  presetsBank;
    //json j( presetsBank);     //alternative 
      
    //check generated json, the string looks OK, see below
    std::cout << j << std::endl;
    //But this throws an error:
    PresetsBank_t myobject = json::parse(j);`

    

The compiler reports the following error on the last line:

../main/json.hpp:5527:67: error: incomplete type 'nlohmann::detail::wide_string_input_helper<nlohmann::detail::iterator_input_adapter<nlohmann::detail::iter_impl<const nlohmann::basic_json<> > >, 16>' used in nested name specifier
         wide_string_input_helper<BaseInputAdapter, T>::fill_buffer(base_adapter, utf8_bytes, utf8_bytes_index, utf8_bytes_filled);

What do I overlook?

NB:

This is the json string before saving
'[[1,[0.0,0.04141732305288315,0.4503936767578125,0.7649605870246887,0.06299212574958801,-1.8740158081054688,440.0,0.5,0.0,0.14960630238056183,1.0325983762741089,0.03937007859349251]],[2,[0.0,0.41842517256736755,0.9537007212638855,1.2053542137145996,0.14173227548599243,-1.4330708980560303,440.0,0.5,0.0,0.22834645211696625,2.4485037326812744,0.14173227548599243]]]'

**Comment by nlohmann** [comment, marked as answer]
```cpp
PresetsBank_t myobject = json::parse(j);`
```

looks wrong. `j` is of type `json` - you do not need to parse it again. Parsing is for creating a `json` value out of some input (e.g., `std::string` or `std::ifstream`).

Do this instead:

```cpp
PresetsBank_t myobject = j;
```

or

```cpp
auto myobject = j.get<PresetsBank_t>();
```

**Comment by goofy2k** [comment]
Thanks! That helps. One must always get used to a new lib......

Op ma 24 jan. 2022 om 09:25 schreef Niels Lohmann ***@***.***
